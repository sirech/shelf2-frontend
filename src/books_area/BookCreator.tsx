import React from 'react'
import { useHistory } from 'react-router-dom'

import CreateButton from 'components/create_button'

const BookCreator = () => {
  const history = useHistory()
  return <CreateButton onClick={() => history.push('/books/new')} />
}

export default BookCreator
