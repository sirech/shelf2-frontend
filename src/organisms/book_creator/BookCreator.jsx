// @flow

import React from 'react'
import { connect } from 'react-redux'
import { createStructuredSelector } from 'reselect'
import { withRouter } from 'react-router-dom'
import { Helmet } from 'react-helmet'

import {
  Alert,
  Button,
  Modal,
  ModalHeader,
  ModalBody,
  ModalFooter,
} from 'reactstrap'

import { openedSelector, validSelector, errorSelector } from './selectors'

import { actionPicker } from 'state'
import { actions } from 'state/modal'

import CreateButton from 'molecules/create_button'
import BookForm from 'organisms/book_form'

type Props = {
  opened: boolean,
  error: boolean,
  valid: boolean,
  history: object,
}

// exported for testing
export class BookCreator extends React.Component<Props> {
  _form: HTMLFormElement
  goToNewBook: Function
  attachForm: Function
  handleClick: Function

  constructor() {
    super()

    this.goToNewBook = this.goToNewBook.bind(this)
    this.attachForm = this.attachForm.bind(this)
    this.handleClick = this.handleClick.bind(this)
  }

  goToNewBook() {
    this.props.history.push('/books/new')
  }

  attachForm(node: HTMLFormElement) {
    this._form = node
  }

  handleClick() {
    this._form.submit()
  }

  showError() {
    if (this.props.error) {
      return <Alert color="danger">Error creating a book</Alert>
    }
  }

  render() {
    const { opened, valid } = this.props

    return (
      <div>
        {opened && (
          <Helmet>
            <title>Add book</title>
          </Helmet>
        )}
        <CreateButton onClick={this.goToNewBook} />
        <Modal isOpen={opened}>
          <ModalHeader goToNewBook={this.goToNewBook}>Add book</ModalHeader>
          <ModalBody>
            {this.showError()}
            <BookForm attachForm={this.attachForm} />
          </ModalBody>
          <ModalFooter>
            <Button
              color="primary"
              disabled={!valid}
              onClick={this.handleClick}
            >
              Create
            </Button>
            <Button color="secondary" onClick={this.goToNewBook}>
              Cancel
            </Button>
          </ModalFooter>
        </Modal>
      </div>
    )
  }
}

export default withRouter(
  connect(
    (state, props) =>
      createStructuredSelector({
        opened: openedSelector,
        error: errorSelector,
        valid: validSelector,
      })(state),
    actionPicker(['modalToggled'])(actions)
  )(BookCreator)
)
