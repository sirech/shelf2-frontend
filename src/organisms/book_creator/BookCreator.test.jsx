import React from 'react'
import { shallow } from 'enzyme'
import toJson from 'enzyme-to-json'

import { BookCreator } from './BookCreator'

describe('components', () => {
  describe('BookCreator', () => {
    it('renders correctly', () => {
      const component = shallow(<BookCreator opened valid />)
      expect(toJson(component)).toMatchSnapshot()
    })
  })
})
