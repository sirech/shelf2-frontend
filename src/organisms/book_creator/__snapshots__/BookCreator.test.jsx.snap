// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components BookCreator renders correctly 1`] = `
<div>
  <HelmetWrapper
    encodeSpecialCharacters={true}
  >
    <title>
      Add book
    </title>
  </HelmetWrapper>
  <CreateButton
    onClick={[Function]}
  />
  <Modal
    autoFocus={true}
    backdrop={true}
    backdropTransitionTimeout={150}
    fade={true}
    isOpen={true}
    keyboard={true}
    modalTransitionTimeout={300}
    zIndex={1050}
  >
    <ModalHeader
      tag="h4"
      toggle={[Function]}
      wrapTag="div"
    >
      Add book
    </ModalHeader>
    <ModalBody
      tag="div"
    >
      <Connect(BookForm)
        attachForm={[Function]}
      />
    </ModalBody>
    <ModalFooter
      tag="div"
    >
      <Button
        color="primary"
        disabled={false}
        onClick={[Function]}
        tag="button"
      >
        Create
      </Button>
      <Button
        color="secondary"
        onClick={[Function]}
        tag="button"
      >
        Cancel
      </Button>
    </ModalFooter>
  </Modal>
</div>
`;
